### PR TITLE
raftstore: fix panic due to stale peer handling snapshot

### LINKED ISSUE #17469: raftstore: TiKV panics due to self was destroyed from meta during receiving snapshot
### What version of TiKV are you using?
v6.5.6
### What operating system and CPU are you using?
`aarch64-unknown-linux-gnu`

### Steps to reproduce
Run a normal cluster. With adding / removing peers.

### What did you expect?
TiKV shouldn't panic.

### What did happened?
TiKV panicked with the following stack:
```
   0: tikv_util::set_panic_hook::{{closure}}
             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/compo nents/tikv_util/src/lib.rs:508:18
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2032 :9
      std::panicking::rust_panic_with_hook
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:692:13
   2: std::panicking::begin_panic_handler::{{closure }}
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:579:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rust/toolchains/n ightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137:18
   4: rust_begin_unwind
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust /library/std/src/panicking.rs:575:5
   5: core::panicking::panic_fmt
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:65:14
   6: core::panicking::panic_ display
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:138:5
   7: core::panicking::panic_str
             at /rust/toolchains/nightly-2022-11-15-aarch 64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:122:5
   8: core::option::expect_failed
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option .rs:1879:5
   9: core::option::Option<T>::expect
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:741:21
      <std::collections::hash::map::HashMap<K,V,S> as core::ops::index::Index<&Q>>::index
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/collections/hash/map.rs:1340:9
      raftstore::store::fsm::peer::PeerFsmDelegat e<EK,ER,T>::check_snapshot
             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/peer.rs:3103:12
  10: raftstore::store::fsm::peer::PeerFsmDelegate<EK,ER,T>::on_ra ft_message
             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/peer.rs:2541:40
      raftstore::store::fsm::peer::PeerFsmDelegate<EK,ER,T>::handle_msgs
      at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/peer.rs:620:37
  11: <raftstore::store::fsm::store::RaftPoller<EK,ER,T> as batch_system::batch::PollHandler<raftstore: :store::fsm::peer::PeerFsm<EK,ER>,raftstore::store::fsm::store::StoreFsm<EK>>>::handle_normal
             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/store.rs:971:9
   12: batch_system::batch::Poller<N,C,Handler>::poll
             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/batch-system/src/batch.rs:430:27
  13: raftstore::store::worker::refresh_config: :PoolController<N,C,H>::increase_by::{{closure}}
             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/worker/refresh_config.rs:78:21
      <std::thread::Builder as ti kv_util::sys::thread::StdThreadBuildWrapper>::spawn_wrapper::{{closure}}
             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/sys/thread.rs:415:23
      std::sys_common::ba cktrace::__rust_begin_short_backtrace
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:121:18
  14: std::thread::Builder::spawn_unchecked_::{{cl osure}}::{{closure}}
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:551:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function:: FnOnce<()>>::call_once
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /rust/toolchains /nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:483:40
      std::panicking::try
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/librar y/std/src/panicking.rs:447:19
      std::panic::catch_unwind
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:137:14
      std::thread::Builder::spawn_uncheck ed_::{{closure}}
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:550:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rus t/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:513:5
  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rust/toolchains/nigh tly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rust/toolchains/nightly-2022-11-15-aarch64 -unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rust/toolchains/nightly-2022-11-15-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/libra ry/std/src/sys/unix/thread.rs:108:17
  16: <unknown>
  17: <unknown>
```

Here is a redacted log about a suspected region: [a-panic-redacted.log](https://github.com/user-attachments/files/16835781/a-panic-redacted.log)

The full log may contain sensitive data hence cannot be uploaded here.

### PR DESCRIPTION
**Pull request by hbisheng**
### What is changed and how it works?

Issue Number: Close #17469

What's Changed:

```commit-message
The commit fixes a panic in TiKV that occurs in a rare scenario that
involves region splits and immediate removal of the new peer.

When a region splits, the new peer on a follower can be created in two 
ways: (1) By receiving a Raft message from the new region 
(`fn maybe_create_peer`) (2) By applying the split operation locally 
(`fn on_ready_split_region`). 

Depending on timing, a new peer might first be created by a Raft 
message and then again when the split is applied. This is a known 
situation. When it happens, the second peer replaces the first, and the 
first peer is dicarded. However, the discarded peer may continue 
processing existing messages, leading to unexpected states.

The panic can be reproduced with the following sequence of events:
1. The first peer is created by a Raft message and is waiting for a 
   Raft snapshot.
2. The second peer (of the same reason) is created by 
   `on_ready_split_region` when the split operation is applied, 
   replacing the first peer and closing its mailbox (as expected).
3. The second peer is immediately removed (a rare situation). This 
   removes the region metadata.
4. The first peer continues processing the Raft snapshot message, 
   expecting the metadata of the region to exist, causing the panic.
```

In addition, [this comment](https://github.com/tikv/tikv/issues/17469#issuecomment-2348345113) gives an analysis of the logs when the panic occurred. 

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Fixed a TiKV panic due to a stale peer handling snapshot when there is a split and an immediate removal of the new peer. 
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -3436,7 +3436,13 @@ where
         }
 
         let mut meta = self.ctx.store_meta.lock().unwrap();
-        if meta.regions[&self.region_id()] != *self.region() {
+        let region_mismatch = match meta.regions.get(&self.region_id()) {
```
Comment (hbisheng) [inline components/raftstore/src/store/fsm/peer.rs:3439]: This change can avoid the panic, but whether it is sufficient needs more careful thoughts. In some cases, panicking is better than letting the process run and do unexpected things.

**Comment by hbisheng** [comment]
cc @gengliqi 
Do you see any risk with the fix? I'm mostly concerned with the possibility that the stale peer may continue to process peer messages and do unexpected things. But since it's an uninitialized peer, it probably can't do much damage?

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs:3439]: I think this fix is quite rational. If this region cannot be found in `meta.regions` or dose not equal to the one in `meta.regions`, it means that another peer has replaced the current peer. Maybe we should add more comments in there.

**Review by gengliqi — APPROVED** [review]

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (hbisheng) [reply components/raftstore/src/store/fsm/peer.rs:3443]: Added some comments for it.

File: tests/failpoints/cases/test_split_region.rs [hunk 2]
```diff
@@ -552,6 +552,90 @@ fn test_split_not_to_split_existing_tombstone_region() {
...
+fn test_snap_handling_after_peer_is_replaced_by_split_and_removed() {
+    // The following diagram represents the final state of the test:
+    //
+    //                    ┌───────────┐  ┌───────────┐  ┌───────────┐
+    //                    │           │  │           │  │           │
+    //      Region 1      │ Peer 1    │  │ Peer 2    │  │ Peer 3    │
+    //      [k2, +∞)      │           │  │           │  │           │
+    // ───────────────────┼───────────┼──┼───────────┼──┼───────────┼──
+    //                    │           │  │           │  │           │
+    //      Region 1000   │ Peer 1001 │  │ Peer 1003 │  │ Peer 1002 │
```
Comment (overvenus) [inline tests/failpoints/cases/test_split_region.rs:565]:
```suggestion
    //      Region 1000   │ Peer 1001 │  │ Peer 1002 │  │ Peer 1003 │
```

I think 1003 is in store3？

File: tests/failpoints/cases/test_split_region.rs [hunk 3]
```diff
@@ -552,6 +552,90 @@ fn test_split_not_to_split_existing_tombstone_region() {
     must_get_equal(&cluster.get_engine(2), b"k1", b"v1");
 }
 
+#[test]
+fn test_snap_handling_after_peer_is_replaced_by_split_and_removed() {
```
Comment (overvenus) [inline tests/failpoints/cases/test_split_region.rs:556]: Could you add another test for handling raft heartbeat, append and request vote messages after old peer is replaced?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: tests/failpoints/cases/test_split_region.rs (hunk 3 above)
Comment (hbisheng) [reply tests/failpoints/cases/test_split_region.rs]: Added such tests. I tried my best to reuse existing code.

File: tests/failpoints/cases/test_split_region.rs (hunk 2 above)
Comment (hbisheng) [reply tests/failpoints/cases/test_split_region.rs:584]: Store 2 actually had peer 1003 when I examined the logs. I don't know why it turned out that way in tests.

File: tests/failpoints/cases/test_split_region.rs (hunk 3 above)
Comment (hbisheng) [reply tests/failpoints/cases/test_split_region.rs]: One thing to note: only the snapshot message test case caused a panic, other tests passed even without the fix.

**Review by overvenus — APPROVED** [review]

**Pull request merged**